### DISCUSSION TITLE
Syncing behavior of joint cloud and local projects can cause users to silently lose the local project when the audio.com version is deleted

### DISCUSSION THREAD

**Discussion by petersampsonaudacity** [audacity/audacity#7176] [Development]
### Bug description
It is possible to save a project to a local disk and then save the same project on Muse's audio.com cloud.  When this happens the two projects are kept in synch, in lock-step, and are totally inter-dependent.

The crunch issue comes if a user deletes the project from audio.com website, in which case the local AUP3 project remains but it will be empty if you try to open it.  The user is not informed that this has happened and will be blissfully unaware, thinking they still have a good local copy, of this until when sometime later they try to open the project and find it empty with all data gone forever,

Part of the issue here is that that is no documentation on the Muse,com Support site about such synched local/cloud projects.  I have added an alert div note in the 3.6 manual to warn of this - but the Manual's entry on cloud saving is minimal, linking instead to the Muse Support site.

==========================================================

Curiously the obverse is **not** true.

If you first save a project to the audio,com Muse cloud and then use Save As to save a local copy, then the two projects are totally independent and are not synched together.

===========================================================

I have not tested what happens if you have such a dual local/cloud synched project and delete the **CloudProjects** folder in the audacity settings folder (which a user may do as part of a total factory reset).

### Steps to reproduce
1. generate a 30 second stereo chirp
2. Save the project to you local disk (I used Desktop to make it simpler to see what's going on)
3. Observe the AUP3 and the temporary SHM and WAL files
4. Now also save the same project to audio.com (Muse's cloud)
5. wait for synching to complete
6. Close Audacity
7. Observe: the temporary WAL and SHM files disappear
8. Go to audio.com
9. Open the project from there
10. Observe: WAL and SHM files appear alongside the AUP3
11. Delete one of the channels from the audio.com cloud project opened at steps 8-9
12. Close Audacity and save the project when prompted
13. Open the local AUP3 project by dragging it on to an Audacity shortcut
14. Observe: it only has one channel.
15. Careful observation shows that both local and cloud projects are open simultaneously
16. **Now comes the acid test**
17. Exit Audacity
18. Go to audio.com and delete the cloud project _(if you can find the delete hidden behind the ellipsis)_
19. Observe: the local AUP3 file remains on your local disk/desktop
20. Open that AUP3 file
21. Observe synching error message
22. click the OK button to dismiss the error dialog
23. Observe: the local project opens **BUT** it is empty

Now at step 19 the user may think "_that's good I've still got my local copy of the project"_ and it may be many weeks later when they move to step 20 and try to open the project only to find that it is empty and lost forever.

Steps 1-15 are good and I suspect by-design, as it keeps the shadow projects in sync (just like using a RAID disk) - but the user gets no indication that this is the case - no documentation on the Muse Support site where cloud saving is documented

### Expected behavior
At Step 4 the user should be warned that the two projects will be synched

At Step 18 the user should be warned that the local project will remain BUT will be empty.
Better would be if they were offered the option to have the local copy deleted as well.

Or maybe at Step 18 Audacity should save the local project in the state it was at steps 15-17 i.e. with good content.

Behaviors and their implications to be properly documented

### Actual behavior
See STRs above

### Audacity Version
3.5.1, latest build for 3.6.0 and alpha master build for 3.7.0

### Operating system
Windows 10 - but assume all OS

### Additional context
This issue was observed as a result of testing after a user posting on the Forum:
https://forum.audacityteam.org/t/all-sound-files-created-are-corrupted/110101

Koz posted this plea for help, to which I responded:
https://forum.audacityteam.org/t/exciting-new-cloud-services/110499

**Comment by dozzzzer** [comment]
Could  only partially reproduce it on macOS: 
- I _did_ receive a syncing error message with 404 error code at step 21
- It _did not_ open a blank project for me at step 23; Audacity remained in the state as if was just launched.

Will check the STRs on Windows later.

**Comment by SteveDaulton** [comment]
> 8. Go to audio.com
> 9. Open the project from there

On Linux, the "Open in Audacity" button on audio.com opens a new tab with the message:


```
The address wasn’t understood

Firefox doesn’t know how to open this address, because one of the following protocols (audacity) isn’t associated with any program or is not allowed in this context.

    You might need to install other software to open this address.
```

I replaced those two steps with:

8. Re-launch Audacity
9. File menu > Open from cloud.


> 12. Close Audacity.

At this point I was prompted to save before close. I selected "Yes"


> 13. Open the local AUP3 project by dragging it on to an Audacity shortcut

That just opens a new empty project for me, so I replaced that step with:

13. "File menu > Open", and select the file from the local disk.


> 23. Observe: the local project opens BUT it is empty

On my system, I get a new empty project. The project on disk fails to open.
The AUP3 file (on the local disk) has a size of 4,489,216 bytes (4.3 MiB).

**Comment by SteveDaulton** [comment]
I found a workaround that worked  with my test project:

24. Make a copy of the project from the local disk, and give it a new, unique name.
25. Audacity "File menu > Open" and select the copy.
26. The project opened, with the state as observed in step 14 (one channel).

On Linux, the project does not appear to have been damaged by its little adventure, though I find it rather disturbing that this bug was not picked up before release. I am strongly of the opinion that end users should not be treated as beta testers, unless they explicitly agree to be part of a beta testing program.

**Comment by petersampsonaudacity** [comment]
Koz raised some interesting questions for me on the Forum thread, which I answered thus.

>My natural action is to save the Primary, Valuable copy to the local machine and the Emergency Backup to the Cloud. Surprise!!

Well yes indeed.

But I would probably save a backup copy before also opening in the the cloud


>What’s the best way to “clean out” your cloud account without causing chaos?

I can't think of a way - as for me cleaning out the cloud account reams out the shadow local copies.

Workaround: I would suggest:
a) backup the synced local/cloud project(s) to a local disk under a (slightly) different name
b) delete the cloud project(s)
c) delete the now empty local project(s) with your file manager.

_I have not tried this._


>Is there any way to break sync once it’s been established?

Not as far as I can see - I see no mechanism provided for that eventuality.


> What happens to sync if the dog chews through your network cable?

An interesting thought (and one easy to simulate - _even without a dog or a chewed cable_.

Though I suspect that Audacity may just hang as the Cloud project always appears to be the dominant active project with the local copy being the slave shadow.  (which is the total opposite of your use case in your first query above).

_I have not tried this either - though I might give this a go later._

**Comment by petersampsonaudacity** [comment]
I have added this to the save Project page in the alpha 3.6 manual - perhaps similar, or more extensive, could be added to the Support GitBook:

![image](https://github.com/audacity/audacity/assets/50205984/19273159-46e5-4145-b75b-1d8a0c5dee10)

**Comment by petersampsonaudacity** [comment]
@dozzzzer @SteveDaulton 
Apologies - at steps 11-12 I should have said that the project be **Saved**  - I have corrected step 12

I tested (these slightly revised STRs) this morning on my MacBook Pro Monterey macOS 12.7.4

I get exact;y the same results as I did on W10 yesterday.

In particular 
a) at step 21 I get the synching error message
b) at Step 23 the project is empty (lost forever for the user)

**Comment by petersampsonaudacity** [comment]
> Koz raised some interesting questions for me on the Forum thread, which I answered thus.
> 
> > My natural action is to save the Primary, Valuable copy to the local machine and the Emergency Backup to the Cloud.

I am  with Koz here.

My natural workflow would be to work on my local machine and local disk - and use the audio.com Muse cloud as an offsite safety backup - frozen in that state and _**NOT**_  synched.

@LWinterberg @Tantacrul 
To that end might it not be useful to introduce two further commands: 
a) Backup Project to Cloud  - which would save  a backup copy in a frozen state un-synched with the local project
b) Restore Project from Cloud - which would copy the project to local storage (and possible force a Save, or maybe offer a Save which could be accepted or declined)

Both of these commands would have potential over-write issues which would have to be considered and dealt with.


================================================================

My other potential use for project storage on audio.com would be the use case where I was planning to work on a collaborative project.  In which case I would probably start right out with a cloud project rather than using local storage.

I suppose for such a use case one might start with local storage and when ready to collaborate move the project to the cloud for collaboration.  But at that stage there should be no need for a synched local copy.  And anyway how would such local synching work when other collaborators were working on the cloud project?

**Comment by dozzzzer** [comment]
We, or rather Audio.com, should probably implement a dialog warning the user that they will lose the data of the locally saved copy of the project when it's deleted on Audio.com.

**Reply by SteveDaulton** [reply]
Perhaps such a dialog should also indicate, (or link to), how to safely delete the cloud copy without losing their work.
